File: samples/java/src/main/java/com/helpshift/ExportIssuesToCsv.java
package com.helpshift;

import com.mashape.unirest.http.HttpResponse;
import com.mashape.unirest.http.Unirest;
import com.mashape.unirest.http.exceptions.UnirestException;
import org.json.*;
import java.io.*;
import java.util.Calendar;
import org.apache.commons.lang3.StringUtils;


public class ExportIssuesToCsv {


    public JSONObject GetIssues (String domain, String api_key, int page, long created_since) {

        try {
            String api_endpoint = "https://api.helpshift.com/v1/" + domain + "/issues?page-size=1000&created_since=" +
                    created_since + "&page=" + page;

            /*
            * Example API with some filters:
                sort-by creation-time, issue state is new, response includes meta
                "https://api.helpshift.com/v1/" + domain +
                "/issues?sort-by=creation-time&state=%5B%22new%22%5D&includes=%5B%22meta%22%5D&page-size=1000&"

            * Please refer the documentation to add filters to GET /issues API as required.
            * */

            HttpResponse<String> response = Unirest.get(api_endpoint)
                    .basicAuth(api_key, "")
                    .asString();

            JSONObject resp_body = new JSONObject(response.getBody());

            if (response.getStatus() != 200) {
                System.out.println ("Something went wrong: " + resp_body);
                System.exit(1);
            }

            return resp_body;

        } catch (UnirestException exception) {
            System.err.println("Exception in making API request: " + exception.getMessage());
            exception.printStackTrace();
            return null;
        }
    }

    public void getAndSaveToCsv(String domain, String api_key, String issues_file_loc, String messages_file_loc,
                                long created_since) {

        try {
            FileWriter issues_file = new FileWriter(issues_file_loc);
            BufferedWriter bw = new BufferedWriter(issues_file);
            PrintWriter pw = new PrintWriter(bw);
            FileWriter messages_file = new FileWriter(messages_file_loc);
            BufferedWriter bwm = new BufferedWriter(messages_file);
            PrintWriter pwm = new PrintWriter(bwm);

            int current_page = 0;
            String [] issue_field_names = {"domain", "id", "title", "app_id", "assignee_name", "tags", "state", "changed_at"};
            JSONArray issue_fields = new JSONArray(issue_field_names);
            String [] message_field_names = {"issue_id", "body", "author", "attachment"};
            JSONArray message_fields = new JSONArray(message_field_names);

            // Creating CSV file headers
            pw.println(StringUtils.join(issue_field_names,','));
            pwm.println(StringUtils.join(message_field_names,','));

            while (true) {
                current_page++;
                JSONObject resp = GetIssues(domain, api_key, current_page, created_since);
                if (resp.getInt("total-pages") < current_page) {
                    System.out.println("Export completed. Total number of API calls: " + (current_page - 1));
                    break;
                }

                /*
                * The response will contain issue details in nested json structure.
                * Below code will flatten the structure by extracting fields as required.
                * It will also flatten the tags array into a string for better readability.
                * The json keys are the names of columns in the output issue CSV file.
                * */

                JSONArray issues = resp.getJSONArray("issues");
                for (int i = 0 ; i < issues.length() ; i++) {

                    JSONObject issue = issues.getJSONObject(i);
                    JSONObject state_data = issue.getJSONObject("state_data");

                    JSONArray tags_array = issue.getJSONArray("tags");
                    StringBuilder tags = new StringBuilder("");
                    for (int k = 0 ; k < tags_array.length() ; k++) {
                        tags = tags.append(tags_array.getString(k)).append(",");
                    }
                    if (tags.length() > 0)
                        tags.deleteCharAt(tags.length() - 1);

                    issues.getJSONObject(i).put("state",  state_data.get("state"));
                    issues.getJSONObject(i).put("changed_at",  state_data.get("changed_at"));
                    issues.getJSONObject(i).put("tags", tags);

                    /*
                    * Destructuring message json and renaming keys to columns of output message CSV file.
                    * */

                    JSONArray messages = issue.getJSONArray("messages");
                    for (int j = 0 ; j < messages.length() ; j++) {
                        JSONObject message = messages.getJSONObject(j);
                        messages.getJSONObject(j).put("issue_id", issue.get("id"));
                        messages.getJSONObject(j).put("author", message.getJSONObject("author").get("name"));
                        if (message.has("attachment")) {
                            messages.getJSONObject(j).put("attachment", message.getJSONObject("attachment").get("file_name"));
                        }
                    }

                    /*
                    * CDL provides support for converting between JSON and comma delimited lists.
                    * Given the column names and array of json objects with columns as keys,
                    * it returns comma delimited text mapped to columns.
                    * For more details, refer https://github.com/vogella/org.json/blob/master/src/org/json/CDL.java
                    * */

                    pwm.print(CDL.toString(message_fields, messages));
                }
                pw.print(CDL.toString(issue_fields, issues));
            }
            pw.flush();
            pwm.flush();
        }
        catch (IOException e) {
            System.err.println("Exception while writing to file: " + e.getMessage());
            e.printStackTrace();
        }
        catch (Exception e) {
            System.err.println("Unexpected Exception: " + e.getMessage());
            e.printStackTrace();
        }
        return;
    }


    public static void main(String args[]) {
        /*
        * POINTS TO NOTE:
        - Messages are exported to a different file for ease of access.
        - Each message row will have its corresponding issue id.
        - The API will fetch 1000 (maximum allowed) issues per call, as specified in the 'api_endpoint' above.
        - Total number of API calls is equal to the number of total pages in the responses.
        - To limit the total size of response, the program takes number of days N as input.
          Issues created in the last N days will be fetched. To get ALL issues, remove the parameter 'created_since'.
        - A few fields of the response json that may not be important have not been considered in this sample.
        - Please refer the documentation for the complete response structure.
         */

        if (args.length != 1) {
            System.out.println("Please enter number of days N. Issues created in the last N days will be retrieved.\n" +
                    "Usage: java -cp <PATH TO JAR> com.helpshift.ExportIssuesToCsv <NO OF DAYS>");
            System.exit(1);
        }

        Calendar c = Calendar.getInstance();
        try {
            int no_of_days = Integer.parseInt(args[0]);
            c.add(Calendar.DATE, (-1 * no_of_days));
        } catch (Exception e) {
            System.out.println("Wrong value given for number of days = " + args[0]);
            System.exit(2);
        }
        long created_since = c.getTimeInMillis();

        String domain = "<DOMAIN>";
        String api_key = "<API_KEY>";
        String issues_file_location = "<ISSUES_FILE_LOCATION>/issues.csv";
        String messages_file_location = "<MESSAGES_FILE_LOCATION>/messages.csv";

        ExportIssuesToCsv exportIssuesToCsv = new ExportIssuesToCsv();
        exportIssuesToCsv.getAndSaveToCsv(domain, api_key, issues_file_location, messages_file_location, created_since);
    }
}

File: samples/python/export_issues_to_csv.py
#!/usr/bin/env python

"""
Usage:

    pip install -r requirements.txt
    python export_issues_to_csv.py

Notes:

- Messages are exported to a different file for ease of access.
- Each message row will have its corresponding issue id.
- The API will fetch 1000 (maximum allowed) issues per call, as specified in the "api_endpoint" below.
- Total number of API calls is equal to the number of total pages in the responses.
- To limit the total size of response, the program takes number of days N as input.
  Issues created in the last N days will be fetched. To get ALL issues, remove the parameter "created_since".
- A few fields of the response json that may not be important have not been considered in this sample.
- Please refer the documentation for the complete response structure.
- Outbound Messages counts include agents and automation messages counts.
- Inbound Messages counts just include end-user messages counts.
"""

from __future__ import absolute_import, division, print_function, unicode_literals

import os

import arrow
import click
import requests
import unicodecsv as csv


ISSUES_FILE_NAME = 'issues.csv'
MESSAGES_FILE_NAME = 'messages.csv'
ISSUES_FIELD_NAMES = ['domain', 'app_id', 'state', 'changed_at', 'assignee_name', 'id', 'title', 'tags',
                      'inbound_messages_count,', 'outbound_messages_count']
MESSAGE_FIELD_NAMES = ['issue_id', 'author', 'body', 'attachment']
ISSUES_API_URL = 'https://api.helpshift.com/v1/{}/issues'


def fetch_issues(api_key, api_params, api_url, current_page):
    api_params['page'] = current_page
    raw_response = requests.get(api_url, params=api_params, auth=(api_key, ''))
    raw_response.raise_for_status()
    return raw_response.json()


def count_inbound_and_outbound_messages(messages):
    inbound_count, outbound_count = 0, 0
    for message in messages:
        if message['origin'] == 'end-user':
            inbound_count += 1
        elif message['origin'] == 'helpshift':
            outbound_count += 1
    return inbound_count, outbound_count


def construct_issues_file_row(issue):
    inbound_count, outbound_count = count_inbound_and_outbound_messages(issue['messages'])
    return dict(issue,
                assignee_name=issue.get('assignee_name', ''),
                title=issue.get('title', ''),
                state=issue['state_data']['state'],
                tags='|'.join(issue['tags']),
                changed_at=issue['state_data']['changed_at'],
                inbound_messages=inbound_count,
                outbound_messages=outbound_count)


def construct_messages_file_row(issue_id, message):
    return dict(issue_id=issue_id,
                body=message['body'],
                author=message['author']['name'],
                attachment=message.get('attachment', {}).get('file_name'))


@click.command()
@click.option('--domain',
              prompt='Domain',
              help='The domain to fetch issues for')
@click.option('--api-key',
              prompt='API key',
              help='The API key for the domain')
@click.option('--days',
              prompt='Number of days',
              default=30,
              type=int,
              help='Last N days to fetch issues for')
@click.option('--output-directory',
              prompt=True,
              default=os.getcwd,
              type=click.Path(exists=True, dir_okay=True, file_okay=False, writable=True, resolve_path=True),
              help='The directory under which we will store CSV files')
def export_issues_to_csv(domain, api_key, days, output_directory):
    timestamp_in_ms = arrow.utcnow().replace(days=-days).timestamp * 1000
    api_url = ISSUES_API_URL.format(domain)
    issues_path = os.path.join(output_directory, ISSUES_FILE_NAME)
    messages_path = os.path.join(output_directory, MESSAGES_FILE_NAME)

    with open(issues_path, 'wb') as issues_file, open(messages_path, 'wb') as messages_file:
        issue_writer = csv.DictWriter(issues_file, fieldnames=ISSUES_FIELD_NAMES, extrasaction='ignore', restval='')
        message_writer = csv.DictWriter(messages_file, fieldnames=MESSAGE_FIELD_NAMES, extrasaction='ignore', restval='')
        issue_writer.writeheader()
        message_writer.writeheader()

        current_page = 0

        # See https://apidocs.helpshift.com/#!/Issues/get_issues for more parameters
        api_params = {
            'created-since': timestamp_in_ms,
            'page-size': 1000
        }

        while True:
            current_page += 1
            response = fetch_issues(api_key, api_params, api_url, current_page)
            click.echo("Fetched page {}/{}".format(current_page, response['total-pages']))

            for issue in response['issues']:
                issue_writer.writerow(construct_issues_file_row(issue))

                for message in issue['messages']:
                    message_writer.writerow(construct_messages_file_row(issue['id'], message))

            if response['total-pages'] == current_page:
                click.echo('Export completed.')
                break


if __name__ == '__main__':
    export_issues_to_csv()
